pkg/generate/cspc.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2020-2021 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generate
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	cstorv1 "github.com/openebs/api/v2/pkg/apis/cstor/v1"
+	"github.com/openebs/api/v2/pkg/apis/openebs.io/v1alpha1"
+	"github.com/openebs/openebsctl/pkg/client"
+	"github.com/openebs/openebsctl/pkg/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// isPoolValid checks if a CStor pool is valid
+func isPoolTypeValid(raid string) bool {
+	if raid == "stripe" || raid == "mirror" || raid == "raidz" || raid == "raidz2" {
+		return true
+	} else {
+		return false
+	}
+}
+
+// Pool calls the generate routine for different cas-types
+func Pool(nodes []string, devs int, raid string) error {
+	c := client.NewK8sClient()
+	if !isPoolTypeValid(strings.ToLower(raid)) {
+		// TODO: Use the well defined pool constant types from openebs/api when added there
+		return fmt.Errorf("invalid pool type %s", raid)
+	}
+	_, str, err := CSPC(c, nodes, devs, strings.ToLower(raid))
+	if err != nil {
+		return err
+	}
+	fmt.Println(str)
+	return nil
+}
+
+// CSPC takes eligible nodes, number of devices and poolType to create a pool cluster template
+func CSPC(c *client.K8sClient, nodes []string, devs int, poolType string) (*cstorv1.CStorPoolCluster, string, error) {
+	// 0. Figure out the OPENEBS_NAMESPACE for CStor
+	cstorNS, err := c.GetOpenEBSNamespace(util.CstorCasType)
+	// assume CSTOR's OPENEBS_NAMESPACE has all the relevant blockdevices
+	c.Ns = cstorNS
+	if err != nil {
+		return nil, "", fmt.Errorf("cannot find an active cstor installation")
+	}
+	// 1. Validate nodes & poolType, fetch disks
+	nodeList, err := c.GetNodes(nodes, "", "")
+	if err != nil {
+		return nil, "", fmt.Errorf("(server error) unable to fetch nodes %s", err)
+	}
+	if len(nodeList.Items) != len(nodes) {
+		return nil, "", fmt.Errorf("not all worker nodes are available for provisioning a CSPC")
+	}
+	// 2. Fetch BD's from the eligible/valid nodes
+	bds, err := c.GetBDs(nil, "kubernetes.io/hostname in ("+strings.Join(nodes, ",")+")")
```

Comment:
This will not work in the case of EKS based clusters since in EKS setup nodeName is different from the `kubernetes.io/hostname` value. So better to get the hostNames for given nodeNames and then query for bds

Reply:
Okay, that shouldn't be >10 lines of code, so I'll do the following for every-case, as it might not be straightforward to figure out if a cluster has this quirk. Found relevant bug @ `https://github.com/kubernetes/kubernetes/issues/54482`

1. Get all nodes by the node-name and then find the hostname and put them in a map[nodeName]HostName.
2. Assuming, `kubectl get nodes` gets me the `metadata.name` which is the nodeName and `metadata.Labels['kubernetes.io/hostname'] will get me the hostName, which I can map and then put them in a comma separated string

Or best, I don't even need a map 🤷🏼